gateway-ha/src/main/java/io/trino/gateway/ha/router/RoutingRulesManager.java

```diff
@@ -0,0 +1,100 @@
+import java.io.IOException;
+import java.nio.channels.FileChannel;
+import java.nio.channels.FileLock;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.util.ArrayList;
+import java.util.List;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+public class RoutingRulesManager
+{
+    private final String rulesConfigPath;
+
+    @Inject
+    public RoutingRulesManager(HaGatewayConfiguration configuration)
+    {
+        this.rulesConfigPath = configuration.getRoutingRules().getRulesConfigPath();
+    }
+
+    public List<RoutingRule> getRoutingRules()
+            throws IOException
+    {
+        ObjectMapper yamlReader = new ObjectMapper(new YAMLFactory());
+        ImmutableList.Builder<RoutingRule> routingRulesBuilder = ImmutableList.builder();
+        try {
+            String content = Files.readString(Paths.get(rulesConfigPath), UTF_8);
+            YAMLParser parser = new YAMLFactory().createParser(content);
+            while (parser.nextToken() != null) {
+                RoutingRule routingRule = yamlReader.readValue(parser, RoutingRule.class);
+                routingRulesBuilder.add(routingRule);
+            }
+            return routingRulesBuilder.build();
+        }
+        catch (IOException e) {
+            throw new IOException("Failed to read or parse routing rules configuration from path : " + rulesConfigPath, e);
+        }
+    }
+
+    public synchronized List<RoutingRule> updateRoutingRule(RoutingRule routingRule)
+            throws IOException
+    {
+        ImmutableList.Builder<RoutingRule> updatedRoutingRulesBuilder = ImmutableList.builder();
+        List<RoutingRule> currentRoutingRulesList = new ArrayList<>();
+        ObjectMapper yamlReader = new ObjectMapper(new YAMLFactory());
+        try {
+            String content = Files.readString(Paths.get(rulesConfigPath), UTF_8);
+            YAMLParser parser = new YAMLFactory().createParser(content);
+            while (parser.nextToken() != null) {
+                RoutingRule currentRoutingRule = yamlReader.readValue(parser, RoutingRule.class);
+                currentRoutingRulesList.add(currentRoutingRule);
+            }
+            for (int i = 0; i < currentRoutingRulesList.size(); i++) {
+                if (currentRoutingRulesList.get(i).name().equals(routingRule.name())) {
+                    currentRoutingRulesList.set(i, routingRule);
+                    break;
+                }
+            }
+            ObjectMapper yamlWriter = new ObjectMapper(new YAMLFactory());
+            StringBuilder yamlContent = new StringBuilder();
+            for (RoutingRule rule : currentRoutingRulesList) {
+                yamlContent.append(yamlWriter.writeValueAsString(rule));
+                updatedRoutingRulesBuilder.add(rule);
+            }
+            try (FileChannel fileChannel = FileChannel.open(Paths.get(rulesConfigPath), StandardOpenOption.WRITE, StandardOpenOption.READ);
```

Comment:
Lock the file before it is read. Otherwise you allow a sequence like
1. Gateway 1 reads rules
2. Gateway 2 reads rules
3. Gateway 1 locks rules, writes rules, releases lock.
4. Gateway 2 locks rules, writes rules, releases lock. 

Which results in the changes by Gateway 1 being lost.